File: 2020/day14/sol.rs
use std::collections::HashMap;
use std::fs;

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let mut memory: HashMap<u64, u64> = HashMap::new();
    let mut mask: String = String::new();
    let mut program: Vec<_> = Vec::new();

    input.lines().by_ref().for_each(|line| {
        if line.starts_with("mask") {
            let (_, value) = line.rsplit_once(' ').unwrap();
            mask = value.to_owned();
        } else {
            let (address, value) = line
                .split_once('[')
                .unwrap()
                .1
                .split_once("] = ")
                .map(|(a, v)| (a.parse::<u64>().unwrap(), v.parse::<u64>().unwrap()))
                .unwrap();

            program.push((address, value, mask.clone()));
            memory.insert(address, apply_mask(value, &mask));
        }
    });

    let memory_sum: u64 = memory.values().sum();

    println!("part 1: {}", memory_sum);

    memory.clear();

    program.iter().for_each(|(address, value, m)| {
        fill_memory(&mut memory, *address, *value, 36, m);
    });

    println!("part 2: {}", memory.values().sum::<u64>());
}

fn apply_mask(value: u64, mask: &str) -> u64 {
    let mut masked_value: u64 = value;

    mask.chars().enumerate().for_each(|(i, c)| match c {
        '0' => masked_value &= !(1 << (35 - i)),
        '1' => masked_value |= 1 << (35 - i),
        _ => {}
    });

    masked_value
}

fn fill_memory(memory: &mut HashMap<u64, u64>, address: u64, value: u64, index: usize, mask: &str) {
    if index == 0 {
        memory.insert(address, value);
        return;
    }

    let next_bit = mask.chars().nth(index - 1).unwrap();

    match next_bit {
        '0' => fill_memory(memory, address, value, index - 1, mask),
        '1' => fill_memory(
            memory,
            address | (1 << (36 - index)),
            value,
            index - 1,
            mask,
        ),
        'X' => {
            fill_memory(
                memory,
                address | (1 << (36 - index)),
                value,
                index - 1,
                mask,
            );
            fill_memory(
                memory,
                address & !(1 << (36 - index)),
                value,
                index - 1,
                mask,
            );
        }
        _ => {}
    }
}

File: 2020/day14/sol.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <vector>

using u64 = uint64_t;
using u32 = uint32_t;

u32 get_address(const std::string& mem) {
    size_t i = 4;
    while (mem[i] != ']') {
        i++;
    }
    return std::stoul(mem.substr(5 - 1, 5 + i));
}

u64 apply_mask(const std::string& mask, const u64 num) {
    u64 masked = num;
    for (size_t i = 0; i < 36; ++i) {
        if (mask[i] != 'X') {
            if (mask[i] == '1') {
                masked |= (1UL << (35 - i));
            } else {
                masked &= ~(1UL << (35 - i));
            }
        }
    }

    return masked;
}

void fill_memory(std::map<u64, u64>& memory, u64 address, u64 value, std::string mask, u32 i = 36) {
    if (i == 0) {
        memory[address] = value;
        return;
    }

    auto next_bit = mask[i - 1];

    switch (next_bit) {
    case '0':
        fill_memory(memory, address, value, mask, i - 1);
        break;
    case '1':
        fill_memory(memory, address |  (1ULL << (36 - i)), value, mask, i - 1);
        break;
    case 'X':
        fill_memory(memory, address |  (1ULL << (36 - i)), value, mask, i - 1);
        fill_memory(memory, address & ~(1ULL << (36 - i)), value, mask, i - 1);
        break;
    default: {}
    }
}

int main() {
    std::ifstream input("input");
    std::map<u64, u64> memory;
    std::vector<std::tuple<u64, u64, std::string>> program;

    std::string mask;
    std::string line;
    while (std::getline(input, line)) {
        std::string var = line.substr(0, line.find(' '));
        std::string arg = line.substr(line.rfind(' ') + 1);

        if (var == "mask") {
            mask = arg;
        } else {
            u32 current_mem = get_address(var);
            u64 value = std::stoull(arg);
            memory[current_mem] = apply_mask(mask, value);
            program.push_back({ current_mem, value, mask });
        }
    }

    u64 sum1 = 0;
    for (const auto& [_, x] : memory) {
        sum1 += x;
    }

    std::cout << "part 1: " << sum1 << '\n';

    memory.clear();

    for (auto [address, value, mask] : program) {
        fill_memory(memory, address, value, mask);
    }

    u64 sum2 = 0;
    for (const auto& [_, x] : memory) {
        sum2 += x;
    }

    std::cout << "part 2: " << sum2 << '\n';

    return 0;
}
